crates/optimism/chainspec/src/lib.rs

```diff
@@ -32,21 +32,53 @@ use reth_chainspec::{
     Hardforks, Head,
 };
 use reth_network_peers::NodeRecord;
+use reth_optimism_forks::OptimismHardfork;
 use reth_primitives_traits::Header;
 
+const DENOMINATOR_MASK: u64 = 0xFFFFFFFF00000000;
+const ELASTICITY_MASK: u64 = 0x00000000FFFFFFFF;
+
 /// OP stack chain spec type.
 #[derive(Debug, Clone, Deref, Into, Constructor, PartialEq, Eq)]
 pub struct OpChainSpec {
     /// [`ChainSpec`].
     pub inner: ChainSpec,
 }
 
+/// Fee trait for OP chain specs.
+pub trait Fee {
+    /// Read from parent to determine the base fee for the next block
+    fn next_block_base_fee(&self, parent: &Header, timestamp: u64) -> U256;
+}
+
 /// Returns the signature for the optimism deposit transactions, which don't include a
 /// signature.
 pub fn optimism_deposit_tx_signature() -> Signature {
     Signature::new(U256::ZERO, U256::ZERO, Parity::Parity(false))
 }
 
+impl Fee for OpChainSpec {
+    fn next_block_base_fee(&self, parent: &Header, timestamp: u64) -> U256 {
```

Comment:
Ah I'm not sure if this will work - we might need to update the spec. I didn't run into this in `kona` because we don't have an equivalent function. Right now the base fee params sent through the attributes are used _within_ the block being executed, meaning looking @ the parent's `nonce` field is not necessarily the base fee params we will use.

Instead, we should probably specify that the params encoded in the parent's `nonce` field will be used, like in this implementation, and we just set the `nonce` field with the attributes' parameters to be used in the _next_ block. That should ensure we can still compute the next block's base fee from just the parent + block timestamp.

Reply:
ref, where kona is doing this differently atm - https://github.com/anton-rs/kona/blob/main/crates/executor/src/lib.rs#L545-L580

I think on our end, we should change this to use the _previous_ block's encoded params (or default, if the previous block's `nonce` field is `0`).

Reply:
https://github.com/ethereum-optimism/specs/pull/418